Migrate to Azul Metadata API

nskobelevs opened the pull request:
### Description

Migrates updater from using the Zulu OpenJDK Discovery API to the Azul Metadata API.

[Migrate from Zulu OpenJDK Discovery API to Azul Metadata API](https://docs.azul.com/core/metadata-api-migration#response-differences)

### Changes
* Rename all references of 'bundle' to 'package' to be in line with API documentation.
* The base URL of the api changed from `https://api.azul.com/zulu/download/community/v1.0/` to `https://api.azul.com/metadata/v1/zulu/packages/`
* `/bundles/latest` is no longer available
  - Now we fetch `/packages/` with `latest=true` param and get the first package in the list (From article: Data are sorted, with the newest packages being returned first.)
* Other param changes:
  - `ext=dmg` -> `archive_type=dmg`
  - `bundle_type=jdk` -> `java_package_type=jdk`
  - `javafx=false` -> `javafx_bundled=false`
  - Added `availability_type=ca` to since `/community/` is no longer part of the path
* Return field changes:
  - `url` -> `download_url`
  - `zulu_version` -> `distro_version`
* Add extra request to `/packages/{package_uuid}` as the `/packages` endpoint does not returns the SHA256 hash.

### Possible Improvements

There are some aspects we could change which I would like some feedback on.

**Consolidate all per-package requests into a single request**
Since `latest` is now a param and `/packages` returns a list we could do an initial request without `java_version` and `arch` to get a list the all the latest packages for all versions and architectures to avoid having to make individual requests. While `/packages` also no longer specifies the architecture we can get the architecture from `/packages/{package_uuid}` which we fetch for the hash anyways.

The only odd thing I noticed is that Java 7 and Java 8 have multiple packages marked as latest and I'm not sure how that should be interpreted. (For example `8.0.372`, `8.0.201` and `8.0.163` each have packages marked as "latest"). We could just use the order of the response as the migration article above does specify that "newest packages being returned first."

This could offset the extra requests we have to make now to get the hash.

**Remove 12.0.2 skip**
I saw the weird quirk with 12.0.2 and checked out the version on the new API and while there are two packages with that version only one of them is marked as "latest" so we could remove skip and just use that one in line with the other versions? (The latest one is zulu12.3.11 without _2)

### Testing
Running the updater after the changes and deleting the casks produces identical `.rb` files as ones present before.

nskobelevs committed:
Migrate to Azul Metadata API

mdogan commented:
@nskobelevs thanks for the PR. 

- Sending a single `/packages` request to fetch all packages sounds fine. But this already runs on GH actions once a few hours, it's not that critical to me.
- Removing `12.0.2` check would be nice if it's resolved now. Old api was returning two different version strings randomly.

nskobelevs committed:
Remove skip for 12.0.2

nskobelevs commented:
I've removed the 12.0.2 check there (`.rb` also updated). I ran a script that made the request every so often over a few hours and the latest package never changed so should be grand but can revert change if issue keeps happening with new API.

I'll skip the request consolidation for now and instead make it part of #50 as I think it'll make more sense then.

You're free to merge if all looks good to you!

nskobelevs committed:
Fix formatting

mdogan commented:
@nskobelevs; we need to revert `12.0.2` version check. See https://github.com/mdogan/homebrew-zulu/pull/52

nskobelevs commented:
Was just about to make to make a PR but just saw your commit there so you beat me to it

This is some really odd behaviour for an API!

mdogan commented:
Yeah, that's weird. I guess both versions have the same order, so api randomly returns one. But version 12 is already EOL, so ignoring is simpler.
